src/tree/tree_manager.rs

```diff
@@ -104,30 +97,21 @@ impl TreeVersion for CanonicalTree {
     fn spawn<M: Middleware + 'static>(
         tx: Sender<Self::ChannelData>,
         block_scanner: Arc<BlockScanner<M>>,
-    ) -> JoinHandle<Result<(), WorldTreeError<M>>> {
+    ) -> JoinHandle<WorldTreeResult<()>> {
         tokio::spawn(async move {
-            let chain_id = block_scanner
-                .middleware
-                .get_chainid()
-                .await
-                .map_err(WorldTreeError::MiddlewareError)?
-                .as_u64();
+            let chain_id =
+                block_scanner.middleware.get_chainid().await?.as_u64();
 
             tracing::info!(?chain_id, "Starting canonical tree manager");
 
-            loop {
-                async {
-                    let (num_blocks, logs) = block_scanner.next().await?;
-
-                    if num_blocks == 0 {
-                        tokio::time::sleep(Duration::from_secs(
-                            BLOCK_SCANNER_SLEEP_TIME,
-                        ))
-                        .await;
-
-                        return ok(());
-                    }
-
+            block_scanner
+                .block_stream()
+                // Retrieve logs concurrently
+                // Setting this too high can cause a 502
+                .buffered(4)
```

Comment:
Will `4` be enough to sync on mainnet?

Reply:
I'm not sure what you mean. The 4 here means we can request blocks 4 times concurrently. Previously this was effectively less than 1.

Reply:
I misread, disregard!